src/config/key_bindings.rs

```diff
@@ -0,0 +1,295 @@
+/// handled by the compositor.
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Hash)]
+#[serde(deny_unknown_fields)]
+pub struct KeyPattern {
+    /// What modifiers are expected to be pressed alongside the key
+    #[serde(deserialize_with = "deserialize_KeyModifiers")]
+    pub modifiers: KeyModifiers,
+    /// The actual key, that was pressed
+    #[serde(deserialize_with = "deserialize_Keysym")]
+    pub key: u32,
+}
+
+impl KeyPattern {
+    pub fn new(modifiers: impl Into<KeyModifiers>, key: u32) -> KeyPattern {
+        KeyPattern {
+            modifiers: modifiers.into(),
+            key,
+        }
+    }
+}
+
+impl ToString for KeyPattern {
+    fn to_string(&self) -> String {
+        let mut result = String::new();
+        if self.modifiers.logo {
+            result += "Super+";
+        }
+        if self.modifiers.ctrl {
+            result += "Ctrl+";
+        }
+        if self.modifiers.alt {
+            result += "Alt+";
+        }
+        if self.modifiers.shift {
+            result += "Shift+";
+        }
+        result += &keysym_get_name(self.key);
+        result
+    }
+}
+
+#[derive(Debug, Deserialize, Clone, PartialEq, Eq)]
+pub enum Action {
+    Terminate,
+    Debug,
+    Close,
+
+    Workspace(u8),
+    NextWorkspace,
+    PreviousWorkspace,
+    LastWorkspace,
+    MoveToWorkspace(u8),
+    MoveToNextWorkspace,
+    MoveToPreviousWorkspace,
+    MoveToLastWorkspace,
+    SendToWorkspace(u8),
+    SendToNextWorkspace,
+    SendToPreviousWorkspace,
+    SendToLastWorkspace,
+
+    NextOutput,
+    PreviousOutput,
+    MoveToNextOutput,
+    MoveToPreviousOutput,
+    SendToNextOutput,
+    SendToPreviousOutput,
+
+    Focus(FocusDirection),
+    Move(Direction),
+
+    ToggleOrientation,
+    Orientation(crate::shell::layout::Orientation),
+
+    ToggleStacking,
+
+    ToggleTiling,
+    ToggleWindowFloating,
+
+    Resizing(ResizeDirection),
+    #[serde(skip)]
+    _ResizingInternal(ResizeDirection, ResizeEdge, KeyState),
+    Maximize,
+    Spawn(String),
+}
+
+fn insert_binding(
+    key_bindings: &mut HashMap<KeyPattern, Action>,
+    modifiers: KeyModifiers,
+    keys: impl Iterator<Item = u32>,
+    action: Action,
+) {
+    if !key_bindings.values().any(|a| a == &action) {
+        for key in keys {
+            let pattern = KeyPattern {
+                modifiers: modifiers.clone(),
+                key,
+            };
+            if !key_bindings.contains_key(&pattern) {
+                key_bindings.insert(pattern, action.clone());
+            }
+        }
+    }
+}
+
+pub fn add_default_bindings(
+    key_bindings: &mut HashMap<KeyPattern, Action>,
+    workspace_layout: WorkspaceLayout,
+) {
+    let (workspace_previous, workspace_next, output_previous, output_next) = match workspace_layout
+    {
+        WorkspaceLayout::Horizontal => (
+            [KeySyms::KEY_Left, KeySyms::KEY_h],
+            [KeySyms::KEY_Right, KeySyms::KEY_j],
```

Comment:
```suggestion
            [KeySyms::KEY_Right, KeySyms::KEY_l],
```

Copy paste issue from my original code